### DISCUSSION TITLE
How to remove Facebook Click Identifier from middleware.ts?

### DISCUSSION THREAD

**Discussion by tuanductran** [vercel/next.js#44824] [Help]
Hello team, how to remove fbclid (Facebook Click Identifier) from middleware.ts?

I have customized the middleware.ts file as follows:

```ts
import { NextRequest, NextResponse } from "next/server";

const allowedParams = ["allowed"];

export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - fbclid (Facebook Click Identifier)
     */
    '/(.*?)([?&]fbclid=[a-zA-Z0-9_-]+)',
  ],
};

export function middleware(req: NextRequest) {
  const url = req.nextUrl;
  let changed = false;

  url.searchParams.forEach((_, key) => {
    if (!allowedParams.includes(key)) {
      url.searchParams.delete(key);
      changed = true;
    }
  });

  // Avoid infinite loop by only redirecting if the query
  // params were changed
  if (changed) {
    return NextResponse.redirect(url);
    // It's also useful to do a rewrite instead of a redirect
    // return NextResponse.rewrite(url)
  }
}
```

When I yarn build and yarn start to test, it doesn't work

**Comment by icyJoseph** [comment]
Mmm I am not sure if the queries are used to match. AFAIK, the config object is digested into AST and from there they match paths statically. 

I guess you have to check in the middleware if the query contains `fclid`,  but yeah, that'll still engage the middleware every time. Perhaps this is best done in next.config.js? redirect based off queries? https://nextjs.org/docs/api-reference/next.config.js/redirects#header-cookie-and-query-matching
